### ISSUE TITLE
impossible to place mushroom stems next to each other with the builder process

### ISSUE THREAD

**Issue by Redhawk18** [cabaletta/baritone#4318]
## How to reproduce
place 3 mushroom stem next to each other, each one will have different nbt tags. Baritone sees the directional nbt tags and tries to find the exact block that matches it. However a `mushroom_stem` isn't counter as a `mushroom_stem[north=false]`

![2024-03-30_21 46 27](https://github.com/cabaletta/baritone/assets/77415970/6f682a3e-3486-4981-a031-8207af5305ff)
![2024-03-30_21 46 30](https://github.com/cabaletta/baritone/assets/77415970/c40a9410-35e2-4385-bf1e-452b48c95b46)

This has to be a problem in the builder process somewhere, I'm not that familiar with it but I wouldn't mind trying to tackle this.

**Comment by ZacSharp** [comment]
The easy solution: `#buildIgnoreProperties north,south,west,east,up,down`

> This has to be a problem in the builder process somewhere, I'm not that familiar with it but I wouldn't mind trying to tackle this.

Baritone uses Minecrafts placement logic to predict what will be placed and if the resulting state doesn't match the required state it won't place. There is no place where it anticipates state changes (e.g. due to placing the next stem) so I guess that's the part you would need to change. (Also the way `approxPlaceable` is used doesn't play well with directional / locational blocks)

**Comment by Redhawk18** [comment]
> The easy solution: `#buildIgnoreProperties north,south,west,east,up,down`
> 
> > This has to be a problem in the builder process somewhere, I'm not that familiar with it but I wouldn't mind trying to tackle this.
> 
> Baritone uses Minecrafts placement logic to predict what will be placed and if the resulting state doesn't match the required state it won't place. There is no place where it anticipates state changes (e.g. due to placing the next stem) so I guess that's the part you would need to change. (Also the way `approxPlaceable` is used doesn't play well with directional / locational blocks)

it feels like bad practice, but ignore those blockstates for only mushroom stems would seem to fix the problem. Also making it easier to support new blocks that act similar with "impossible" states baritone cannot place.

**Comment by Ali90922** [comment]
Baritone's having a rough time with those mushroom stems and their different NBT tags, huh? It's messing up the whole direction thing. Sure, ignoring some block properties might patch things up for now, but we gotta get down to the nitty-gritty and find a real fix

**Comment by ZacSharp** [comment]
> it feels like bad practice, but ignore those blockstates for only mushroom stems would seem to fix the problem

Yeah, when I made the setting I considered making it per-block but since I didn't anticipate how useful it would be I thought it would be "overkill". Changing `buildIgnoreProperties` per block in a backwards incompatible manner is easy, as is adding a new setting. Do you think once there is a per-block setting there will be much use for the current `buildIgnoreProperties`?


@Ali90922 if you don't even know the problem, can you please not tell others to fix it? (mushroom stems have no nbt attached to them)

**Comment by ZacSharp** [comment]
Hardcoding some cases like "ignore north for mushroom_stem if there will be a block north of it" would also be a simple short term solution, but I don't like it at all. Currently validity checking is almost completely independent of the rest of the builder and this simple fix would add a weird edge case to accommodate a quirk in other parts of the implementation.
(the only thing you could currently consider an entanglement is the `itemCheck` parameter switching between "valid to keep there" and "valid to place there")

**Comment by Redhawk18** [comment]
> Hardcoding some cases like "ignore north for mushroom_stem if there will be a block north of it" would also be a simple short term solution, but I don't like it at all. Currently validity checking is almost completely independent of the rest of the builder and this simple fix would add a weird edge case to accommodate a quirk in other parts of the implementation. (the only thing you could currently consider an entanglement is the `itemCheck` parameter switching between "valid to keep there" and "valid to place there")

could you expand on this?

**Comment by ZacSharp** [comment]
It's just further narrowing down the cases in which the property is ignored and having the behavior hardcoded rather than as a default configuration value.

**Comment by Redhawk18** [comment]
> It's just further narrowing down the cases in which the property is ignored and having the behavior hardcoded rather than as a default configuration value.

oh like manually doing the checks minecraft currently does for mushroom stems and blocks but with the selection or schematic baritone has load?

**Comment by ZacSharp** [comment]
> oh like manually doing the checks minecraft currently does for mushroom stems and blocks but with the selection or schematic baritone has load?

Sort of? The intention would be checking whether the state will be corrected by placing the surrounding blocks and that's pretty similar.
Also I still don't think this is a good idea.

**Comment by Redhawk18** [comment]
recently I was trying to use baritone to build redstone and i found that I have the same issue here, a lot of the blocks blockstate are powered in the schematic but no redstone besides power sources are powered when place by themselves. Another use for this idea
